lib/tenstorrent/bh_arc/eth.c

```diff
@@ -286,76 +297,108 @@ static void SerdesEthInit(void)
+
+	/* Load fw regs */
 	for (uint8_t serdes_inst = 0; serdes_inst < 6; serdes_inst++) {
 		if (load_serdes & (1 << serdes_inst)) {
-			LoadSerdesEthRegs(serdes_inst, ring, (SerdesRegData *)large_sram_buffer,
-					  reg_table_size / sizeof(SerdesRegData));
+			LoadSerdesEthRegs(serdes_inst, ring, buf, spi_address, image_size);
 		}
 	}
 
-	/* Load fw */
-	size_t fw_size = 0;
-
-	if (tt_boot_fs_get_file(&boot_fs_data, ETH_SD_FW_TAG, large_sram_buffer, SCRATCHPAD_SIZE,
-				&fw_size) != TT_BOOT_FS_OK) {
-		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_get_file", ETH_SD_FW_TAG, -EIO);
+	rc = tt_boot_fs_find_fd_by_tag(flash, ETH_SD_FW_TAG, &tag_fd);
+	if (rc < 0) {
+		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_find_fd_by_tag", ETH_SD_FW_TAG, rc);
 		return;
 	}
+	image_size = tag_fd.flags.f.image_size;
+	spi_address = tag_fd.spi_addr;
 
+	/* Load fw */
 	for (uint8_t serdes_inst = 0; serdes_inst < 6; serdes_inst++) {
 		if (load_serdes & (1 << serdes_inst)) {
-			LoadSerdesEthFw(serdes_inst, ring, large_sram_buffer, fw_size);
+			LoadSerdesEthFw(serdes_inst, ring, buf, spi_address, image_size);
 		}
 	}
 }
 
 static void EthInit(void)
 {
 	uint32_t ring = 0;
+	int rc;
+	tt_boot_fs_fd tag_fd;
+	size_t image_size;
+	size_t spi_address;
 
 	/* Early exit if no ETH tiles enabled */
 	if (tile_enable.eth_enabled == 0) {
 		return;
 	}
 
-	/* Load fw */
-	size_t fw_size = 0;
+	uint8_t *buf = (uint8_t *)k_aligned_alloc(4, TEMP_SPI_BUFFER_SIZE);
+
+	if (buf == NULL) {
+		LOG_ERR("%s(%d) failed: %d", "k_aligned_alloc", TEMP_SPI_BUFFER_SIZE, -ENOMEM);
+		return;
+	}
 
-	if (tt_boot_fs_get_file(&boot_fs_data, ETH_FW_TAG, large_sram_buffer, SCRATCHPAD_SIZE,
-				&fw_size) != TT_BOOT_FS_OK) {
-		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_get_file", ETH_FW_TAG, -EIO);
+	rc = tt_boot_fs_find_fd_by_tag(flash, ETH_FW_TAG, &tag_fd);
+	if (rc < 0) {
+		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_find_fd_by_tag", ETH_FW_TAG, rc);
 		return;
 	}
+	image_size = tag_fd.flags.f.image_size;
+	spi_address = tag_fd.spi_addr;
 
+	/* Load fw */
 	for (uint8_t eth_inst = 0; eth_inst < MAX_ETH_INSTANCES; eth_inst++) {
 		if (tile_enable.eth_enabled & BIT(eth_inst)) {
-			LoadEthFw(eth_inst, ring, large_sram_buffer, fw_size);
+			LoadEthFw(eth_inst, ring, buf, spi_address, image_size);
 		}
 	}
 
-	/* Load param table */
-	if (tt_boot_fs_get_file(&boot_fs_data, ETH_FW_CFG_TAG, large_sram_buffer, SCRATCHPAD_SIZE,
-				&fw_size) != TT_BOOT_FS_OK) {
-		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_get_file", ETH_FW_CFG_TAG, -EIO);
+	k_free(buf);
+	buf = NULL;
+
+	rc = tt_boot_fs_find_fd_by_tag(flash, ETH_FW_CFG_TAG, &tag_fd);
+	if (rc < 0) {
+		LOG_ERR("%s(%s) failed: %d", "tt_boot_fs_find_fd_by_tag", ETH_FW_CFG_TAG, rc);
 		return;
 	}
+	image_size = tag_fd.flags.f.image_size;
+	spi_address = tag_fd.spi_addr;
 
+	buf = (uint8_t *)k_aligned_alloc(4, image_size);
+
+	if (buf == NULL) {
+		LOG_ERR("%s(%d) failed: %d", "k_aligned_alloc", image_size, -ENOMEM);
+		return;
+	}
```

Comment:
Can we drop this allocation and just use the previous one?

The main point of this change is to use a buffer that is significantly smaller than the whole image, so that we can DMA to ERISC, MRISC, etc, incrementally rather than all at once.

If necessary, add e.g.

```cpp
	__ASSERT(TEMP_SPI_BUFFER_SIZE >= image_size, "spi buffer size %zu must be larger than image size %zu", TEMP_SPI_BUFFER_SIZE, image_size);
```